packages/api/src/beacon/client/events.ts

```diff
@@ -1,56 +1,52 @@
       const query = stringifyQuery({topics});
-      const url = `${urlJoin(baseUrl, routesData.eventstream.url)}?${query}`;
+      const url = `${urlJoin(baseUrl, definitions.eventstream.url)}?${query}`;
       // eslint-disable-next-line @typescript-eslint/naming-convention
       const EventSource = await getEventSource();
       const eventSource = new EventSource(url);
 
-      try {
-        await new Promise<void>((resolve, reject) => {
-          for (const topic of topics) {
-            eventSource.addEventListener(topic, ((event: MessageEvent) => {
-              const message = eventSerdes.fromJson(topic, JSON.parse(event.data));
-              onEvent({type: topic, message} as BeaconEvent);
-            }) as EventListener);
-          }
-
-          // EventSource will try to reconnect always on all errors
-          // `eventSource.onerror` events are informative but don't indicate the EventSource closed
-          // The only way to abort the connection from the client is via eventSource.close()
-          eventSource.onerror = function onerror(err) {
-            const errEs = err as unknown as EventSourceError;
-            // Consider 400 and 500 status errors unrecoverable, close the eventsource
-            if (errEs.status === 400) {
-              reject(Error(`400 Invalid topics: ${errEs.message}`));
-            }
-            if (errEs.status === 500) {
-              reject(Error(`500 Internal Server Error: ${errEs.message}`));
-            }
-
-            // TODO: else log the error somewhere
-            // console.log("eventstream client error", errEs);
-          };
-
-          // And abort resolve the promise so finally {} eventSource.close()
-          signal.addEventListener("abort", () => resolve(), {once: true});
-        });
-      } finally {
+      const close = (): void => {
         eventSource.close();
+        onClose?.();
+        signal.removeEventListener("abort", close);
+      };
+      signal.addEventListener("abort", close, {once: true});
+
+      for (const topic of topics) {
+        eventSource.addEventListener(topic, (event: MessageEvent) => {
+          const message = eventSerdes.fromJson(topic, JSON.parse(event.data));
+          onEvent({type: topic, message} as BeaconEvent);
+        });
       }
 
-      return {ok: true, response: undefined, status: HttpStatusCode.OK};
+      // EventSource will try to reconnect always on all errors
+      // `eventSource.onerror` events are informative but don't indicate the EventSource closed
+      // The only way to abort the connection from the client is via eventSource.close()
+      eventSource.onerror = function onerror(err) {
+        const errEs = err as unknown as EventSourceError;
+        onError?.(new Error(errEs.message));
+        // Consider 400 and 500 status errors unrecoverable, close the eventsource
+        if (errEs.status === 400 || errEs.status === 500) {
```

Comment:
400 are clients side failures, so these should be recoverable and we should not close stream for it.

Reply:
400 in this context means an invalid topic was passed, this is unrecoverable as the eventsource will never serve any data

This behavior was not changed in this PR but matches how we handle it on unstable branch
https://github.com/ChainSafe/lodestar/blob/3c5bb150c5df9a1dc6fc47462f2d6e20f6054413/packages/api/src/beacon/client/events.ts#L34-L35